src/shared/Microsoft.Git.CredentialManager/HostProviderRegistry.cs

```diff
@@ -46,14 +55,79 @@ public void Register(params IHostProvider[] hostProviders)
+            {
+                throw new ArgumentException(
+                    $"A host provider cannot be registered with the legacy authority ID '{Constants.AuthorityIdAuto}'",
+                    nameof(hostProviders));
+            }
+
             _hostProviders.AddRange(hostProviders);
         }
 
         public IHostProvider GetProvider(InputArguments input)
         {
-            var provider = _hostProviders.FirstOrDefault(x => x.IsSupported(input));
+            IHostProvider provider;
+
+            //
+            // Try and locate a specified provider
+            //
+            if (_context.Settings.ProviderOverride is string providerId)
+            {
+                _context.Trace.WriteLine($"Host provider override was set id='{providerId}'");
+
+                if (!StringComparer.OrdinalIgnoreCase.Equals(Constants.ProviderIdAuto, providerId))
+                {
+                    provider = _hostProviders.FirstOrDefault(x => StringComparer.OrdinalIgnoreCase.Equals(x.Id, providerId));
+
+                    if (provider is null)
+                    {
+                        _context.Trace.WriteLine($"No host provider was found with ID '{providerId}'.. falling back to auto-detection.");
+                    }
+                    else
+                    {
+                        return provider;
+                    }
+                }
+            }
+            //
+            // Try and locate a provider by supported authorities
+            //
+            else if (_context.Settings.LegacyAuthorityOverride is string authority)
+            {
+                _context.Trace.WriteLine($"Host provider authority override was set authority='{authority}'");
+                _context.StdError.WriteLine("warning: the `credential.authority` and `GCM_AUTHORITY` settings are deprecated.");
+                _context.StdError.WriteLine($"warning: see {Constants.HelpUrls.GcmAuthorityDeprecated} for more information.");
+
+                if (!StringComparer.OrdinalIgnoreCase.Equals(Constants.AuthorityIdAuto, authority))
+                {
+                    provider = _hostProviders.FirstOrDefault(x => x.SupportedAuthorityIds.Contains(authority, StringComparer.OrdinalIgnoreCase));
+
+                    if (provider is null)
+                    {
+                        _context.Trace.WriteLine($"No host provider was found with authority '{authority}'.. falling back to auto-detection.");
+                    }
+                    else
+                    {
+                        return provider;
+                    }
+                }
+            }
+
+            //
+            // Auto-detection
+            //
+            _context.Trace.WriteLine("Performing auto-detection of host provider.");
+            provider = _hostProviders.FirstOrDefault(x => x.IsSupported(input));
 
-            if (provider == null)
+            if (provider is null)
```

Comment:
nit: Why "is"?

Reply:
I've been writing GCM Core with the latest language features where possible. I missed this one.